source/common/network/utility.cc

```diff
@@ -292,15 +293,29 @@ Address::InstanceConstSharedPtr Utility::getOriginalDst(int fd) {
 #ifdef SOL_IP
   sockaddr_storage orig_addr;
   socklen_t addr_len = sizeof(sockaddr_storage);
-  int status = getsockopt(fd, SOL_IP, SO_ORIGINAL_DST, &orig_addr, &addr_len);
+
+  int status;
+  switch (orig_addr.ss_family) {
+  case AF_INET:
+    status = getsockopt(fd, SOL_IP, SO_ORIGINAL_DST, &orig_addr, &addr_len);
+  case AF_INET6:
+    status = getsockopt(fd, SOL_IPV6, IP6T_SO_ORIGINAL_DST, &orig_addr, &addr_len);
+  }
 
   if (status == 0) {
-    // TODO(mattklein123): IPv6 support. See github issue #1094.
-    ASSERT(orig_addr.ss_family == AF_INET);
-    return Address::InstanceConstSharedPtr{
-        new Address::Ipv4Instance(reinterpret_cast<sockaddr_in*>(&orig_addr))};
-  } else {
-    return nullptr;
+    if
+      orig_addr.ss_family == AF_INET {
+        return Address::InstanceConstSharedPtr{
+            new Address::Ipv4Instance(reinterpret_cast<sockaddr_in*>(&orig_addr))};
+      }
+    else if
+      orig_addr.ss_family == AF_INET6 {
+        return Address::InstanceConstSharedPtr{
+            new Address::Ipv6Instance(reinterpret_cast<sockaddr_in*>(&orig_addr))};
```

Comment:
the constructor of `Address::Ipv6Instance` has different parameters from `Address::Ipv4Instance`, using `sockaddr_in6&` instead